src/stirling/bpf_tools/rr/rr_bpf_test.cc

```diff
@@ -0,0 +1,179 @@
+NO_OPT_ATTR uint32_t Foo(const uint32_t arg) { return 1 + arg; }
+NO_OPT_ATTR uint32_t Bar(const uint32_t arg) { return 2 + arg; }
+
+std::vector<int> gold_data;
+uint32_t test_idx = 0;
+
+void PerfBufferRecordingDataFn(void* cb_cookie, void* data, int data_size) {
+  DCHECK(cb_cookie != nullptr) << "Perf buffer callback not set-up properly. Missing cb_cookie.";
+  EXPECT_EQ(data_size, sizeof(int));
+  int v = *static_cast<int*>(data);
+  gold_data.push_back(v);
+}
+
+void PerfBufferReplayingDataFn(void* cb_cookie, void* data, int data_size) {
+  DCHECK(cb_cookie != nullptr) << "Perf buffer callback not set-up properly. Missing cb_cookie.";
+  EXPECT_EQ(data_size, sizeof(int));
+  int v = *static_cast<int*>(data);
+  EXPECT_EQ(gold_data[test_idx], v);
+  ++test_idx;
+}
+
+void PerfBufferLossFn(void* cb_cookie, uint64_t /*lost*/) {
+  DCHECK(cb_cookie != nullptr) << "Perf buffer callback not set-up properly. Missing cb_cookie.";
+}
+
+}  // namespace test
+
+namespace px {
+namespace stirling {
+
+using bpf_tools::BPFProbeAttachType;
+using bpf_tools::UProbeSpec;
+using bpf_tools::WrappedBCCArrayTable;
+using bpf_tools::WrappedBCCMap;
+using bpf_tools::WrappedBCCStackTable;
+
+class BasicRecorderTest : public ::testing::Test {
+ public:
+  BasicRecorderTest() {}
+
+ protected:
+  void SetUp() override {
+    recording_bcc_ = std::make_unique<bpf_tools::RecordingBCCWrapperImpl>();
+    replaying_bcc_ = std::make_unique<bpf_tools::ReplayingBCCWrapperImpl>();
+
+    // Register our BPF program in the kernel, for real (recording), and for fake (replaying).
+    ASSERT_OK(recording_bcc_->InitBPFProgram(rr_test_bcc_script));
+    ASSERT_OK(replaying_bcc_->InitBPFProgram(rr_test_bcc_script));
+
+    const auto recording_perf_buffer_specs = MakeArray<bpf_tools::PerfBufferSpec>({
+        .name = std::string("perf_buffer"),
+        .probe_output_fn = test::PerfBufferRecordingDataFn,
+        .probe_loss_fn = test::PerfBufferLossFn,
+        .cb_cookie = this,
+    });
+    const auto replaying_perf_buffer_specs = MakeArray<bpf_tools::PerfBufferSpec>({
+        .name = std::string("perf_buffer"),
+        .probe_output_fn = test::PerfBufferReplayingDataFn,
+        .probe_loss_fn = test::PerfBufferLossFn,
+        .cb_cookie = this,
+    });
+
+    // Open perf buffers for real (recording), and for fake (replaying).
+    ASSERT_OK(recording_bcc_->OpenPerfBuffers(recording_perf_buffer_specs));
+    ASSERT_OK(replaying_bcc_->OpenPerfBuffers(replaying_perf_buffer_specs));
+
+    const int64_t self_pid = getpid();
+    const std::filesystem::path self_path = GetSelfPath().ValueOrDie();
+    ASSERT_OK_AND_ASSIGN(auto elf_reader, ElfReader::Create(self_path.string()));
+    ASSERT_OK_AND_ASSIGN(auto converter, ElfAddressConverter::Create(elf_reader.get(), self_pid));
+
+    // For the uprobe spec.
+    const uint64_t foo_virt_addr = reinterpret_cast<uint64_t>(&::test::Foo);
+    const uint64_t bar_virt_addr = reinterpret_cast<uint64_t>(&::test::Bar);
```

Comment:
In another place where we follow this similar pattern (probing a symbol within the current executable), we define c linkage ([source](https://github.com/pixie-io/pixie/blob/1e0654e33edaab4978b4b4e419f1e19d178aeeb0/src/stirling/source_connectors/socket_tracer/socket_trace_bpf_tables.cc#L36-L42)). It seems like that must be unnecessary since you were able to avoid declaring that? Do you know why that was done in the previous case?

Reply:
I think the c-linkage is not needed in the source code you cite.